### PR TITLE
feat(tags): http

### PR DESCRIPTION
**Pull request by FrankSpitulski**
flavour, user agent, client ip

I had to make a new type of tracer method that takes both the request and the connection because netty has the flavour on the request while armeria has the flavour on the connection. Also, I needed a fallback for the client ip which relies on the connection.

Closes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/776

### CODE REVIEW

**Comment by FrankSpitulski** [comment]
fixing tests, but I figured I'd put it up here for people to review

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 1]
```diff
@@ -195,6 +201,44 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+  protected void onConnectionAndRequest(Span span, CONNECTION connection, REQUEST request) {
+    String flavor = flavor(connection, request);
+    if (flavor != null) {
+      SemanticAttributes.HTTP_FLAVOR.set(span, flavor);
+    }
+    SemanticAttributes.HTTP_CLIENT_IP.set(span, clientIP(connection, request));
+  }
+
+  protected String clientIP(CONNECTION connection, REQUEST request) {
+    // try Forwarded
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:213]: do you think we need to support this header?

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 2]
```diff
@@ -195,6 +201,44 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+          return split[1];
+        }
+      }
+    }
+
+    // try X-Forwarded-For
+    forwarded = requestHeader(request, "X-Forwarded-For");
+    if (forwarded != null) {
+      // may be split by ,
+      String[] split = forwarded.split(", *", 2); // only need the first
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:232]: should be slightly faster to split by `,` and then trim result

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/typedspan/DelegatingSpan.java [hunk 3]
```diff
@@ -91,6 +91,11 @@ public void setStatus(final Status status) {
     delegate.setStatus(status);
   }
 
+  @Override
+  public void recordException(Throwable throwable) {
+    delegate.recordException(throwable);
+  }
+
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/typedspan/DelegatingSpan.java:98]: ya, fix should be merged shortly in #824 👍

File: instrumentation/netty/netty-3.8/src/main/java/io/opentelemetry/auto/instrumentation/netty/v3_8/server/NettyHttpServerTracer.java [hunk 4]
```diff
@@ -67,6 +72,11 @@ protected String peerHostIP(final Channel channel) {
     return null;
   }
 
+  @Override
+  protected String flavor(Channel channel, HttpRequest request) {
+    return null; // TODO flavor is on the request, not the connection for this tracer
```
Comment (trask) [inline instrumentation/netty/netty-3.8/src/main/java/io/opentelemetry/auto/instrumentation/netty/v3_8/server/NettyHttpServerTracer.java:77]: does this work?
```suggestion
    return request.getProtocolVersion().toString();
```

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 5]
```diff
@@ -195,6 +201,44 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+    }
+    SemanticAttributes.HTTP_CLIENT_IP.set(span, clientIP(connection, request));
+  }
+
+  protected String clientIP(CONNECTION connection, REQUEST request) {
+    // try Forwarded
+    String forwarded = requestHeader(request, "Forwarded");
+    if (forwarded != null) {
+      // may be split by ; or ,
+      for (String forwardedValues : forwarded.split(";|(, *)")) {
```
Comment (anuraaga) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:217]: `split(regex)` is really slow, if we need a regex we should use `Pattern.compile` at least. But it'd be best if we can avoid split too due to the array allocations, do you think we can loop through to find the indexes instead of splitting?

@trask By the way, is it ok to use Guava here? `CharMatcher` can help with such searches.

Comment (trask) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:217]: > is it ok to use Guava here?

Currently we don't, maybe better not to add this to our instrumentation api contract. We could expose something in our instrumentation api that wraps specific guava functionality if we need. I'd probably lean towards doing the index looping here for now (assuming we need to support the "Forwarded" header).

Comment (anuraaga) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:217]: Ah forgot this is the instrumentation API, also used in library instrumentation. Agree not to add dependencies.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java (hunk 1 above)
Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:213]: yeah, mozilla says it's the real standard while x-forwarded-for is the defacto standard. best to get both imo. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded

File: instrumentation/netty/netty-3.8/src/main/java/io/opentelemetry/auto/instrumentation/netty/v3_8/server/NettyHttpServerTracer.java (hunk 4 above)
Comment (FrankSpitulski) [reply instrumentation/netty/netty-3.8/src/main/java/io/opentelemetry/auto/instrumentation/netty/v3_8/server/NettyHttpServerTracer.java:77]: yep, I just missed this one when I added the request param.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java (hunk 5 above)
Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:217]: I'll change it to pattern.compile and incorporate the trim optimisation

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java (hunk 2 above)
Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java]: I got rid of the regex in the x-forwarded-for case

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/typedspan/DelegatingSpan.java (hunk 3 above)
Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/typedspan/DelegatingSpan.java]: rebased

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 6]
```diff
@@ -195,6 +205,48 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+      }
+    }
+
+    // try X-Forwarded-For
+    forwarded = requestHeader(request, "X-Forwarded-For");
+    if (forwarded != null) {
+      // may be split by ,
+      int endIndex = forwarded.indexOf(',');
+      if (endIndex > 0) {
+        forwarded = forwarded.substring(0, endIndex);
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:239]: did we need trim here?
```suggestion
        forwarded = forwarded.substring(0, endIndex).trim();
```

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 7]
```diff
@@ -195,6 +205,48 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+  }
+
+  protected String clientIP(CONNECTION connection, REQUEST request) {
+    // try Forwarded
+    String forwarded = requestHeader(request, "Forwarded");
+    if (forwarded != null) {
+      // may be split by ; or ,
+      for (String forwardedValues : FORWARDED_SPLIT_PATTERN.split(forwarded)) {
+        forwardedValues = forwardedValues.trim();
+        String[] split = EQUALS_PATTERN.split(forwardedValues, 2);
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:223]: can you convert `EQUALS_PATTERN` to indexOf/substring to avoid that regex? I don't think that would add code complexity.

replacing `FORWARDED_SPLIT_PATTERN` would definitely add code complexity, so i don't mind keeping

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java (hunk 6 above)
Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:239]: no, there aren't going to be any spaces before the first comma and we only need the first value

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java (hunk 7 above)
Comment (anuraaga) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:223]: Out of curiosity, what do you think about this? I noticed the lines of code are very close to what we have right now even though we define a helper method.

```
int forIndex = forwarded.indexOf("for=");
int endIndex = findForwardedEnd(forwarded, forIndex + 4);
forwarded.substring(forIndex + 4, endIndex);

int find forwardedEnd(String s, int start) {
  for (int i = start; i < s.length(); i++) {
    char c = s.charAt(i);
    if (c == ',' || c == ';') {
      return i;
    }
  }
  return s.length()
```

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java]: @anuraaga this would work for regular cases but would not catch a malformed header such as "for=, for=1.1.1.1" That's probably okay though. I will use your implementation.

**Comment by FrankSpitulski** [comment]
@trask I think this may be a sporadic test failure, works on my machine and it's a timeout error 🤷 

https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-java-instrumentation/2264/workflows/139f7258-b674-4f8f-abfc-87311eecdd25/jobs/14878/tests

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 8]
```diff
@@ -195,6 +201,59 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+      if (!forwarded.isEmpty()) {
+        return forwarded;
+      }
+    }
+
+    // fallback to peer IP if there are no proxy headers
+    return peerHostIP(connection);
+  }
+
+  private static String extractForwardedFor(String forwarded) {
```
Comment (trask) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:239]: Can you add unit test for this method (to mitigate against the extra complexity we asked for :smile:)?

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java]: added tests, caught an edge case

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java]: getting another issue with latestDepTest though.

./gradlew :instrumentation:grpc-1.5:latestDepTest fails
./gradlew :instrumentation:grpc-1.5:test works

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java [hunk 9]
```diff
@@ -195,6 +201,63 @@ protected void onRequest(final Span span, final REQUEST request) {
...
+      }
+    }
+
+    // fallback to peer IP if there are no proxy headers
+    return peerHostIP(connection);
+  }
+
+  // VisibleForTesting
+  static String extractForwardedFor(String forwarded) {
+    int start = forwarded.toLowerCase().indexOf("for=");
```
Comment (anuraaga) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:241]: Since the spec seems to use lowercase, and it's not a commonly used header anyways, I probably wouldn't be so permissive and just not do the `toLowerCase` but it's ok too. From what I can tell, we consider it unimportant enough to not have unit tests :P

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:241]: the mozilla spec says it can be any case, so I just stuck that in there. I've never seen it used in practice but it's apparently the standard.

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:241]: there's an uppercase example in the RFC https://tools.ietf.org/html/rfc7239#section-4

Comment (anuraaga) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:241]: Ah thanks I couldn't find it, in that case can we add one test case for it?

Comment (FrankSpitulski) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java:241]: sure, added

**Comment by trask** [comment]
I forgot that I can merge master into your branch 😄
